cub/device/dispatch/dispatch_radix_sort.cuh

```diff
@@ -1352,33 +1374,145 @@ struct DispatchRadixSort :
+                                    part < num_parts - 1 ?
+                                        d_bins + ((part + 1) * num_passes + pass) * RADIX_DIGITS : NULL,
+                                    d_bins + (part * num_passes + pass) * RADIX_DIGITS,
+                                    d_keys_out_,
+                                    d_keys_in_ + part * PART_SIZE,
+                                    d_values_out_,
+                                    d_values_in_ + part * PART_SIZE,
+                                    part_num_items, current_bit, num_bits);
+                                break;
+                            }
+                            case OUTPUT: {
+                                using KeyInIterT = KeyIteratorT;
+                                using ValueInIterT = ValueIteratorT;
+                                auto onesweep_kernel = DeviceRadixSortOnesweepKernel<
+                                    MaxPolicyT, IS_DESCENDING, KeyInIterT, KeyOutIterT,
+                                    ValueInIterT, ValueOutIterT, OffsetT>;
+                                KeyInIterT d_keys_in_ = d_keys_out;
+                                ValueInIterT d_values_in_ = d_values_out;
+                                onesweep_kernel<<<num_blocks, ONESWEEP_BLOCK_THREADS, 0, stream>>>
+                                    (d_lookback, d_ctrs + part * num_passes + pass,
+                                    part < num_parts - 1 ?
+                                        d_bins + ((part + 1) * num_passes + pass) * RADIX_DIGITS : NULL,
+                                    d_bins + (part * num_passes + pass) * RADIX_DIGITS,
+                                    d_keys_out_,
+                                    d_keys_in_ + part * PART_SIZE,
+                                    d_values_out_,
+                                    d_values_in_ + part * PART_SIZE,
+                                    part_num_items, current_bit, num_bits);
+                                break;
+                            }
+                        }
+                    }
                     if (CubDebug(error = cudaPeekAtLastError())) break;
                 }
-                
-                // use the temporary buffers if no overwrite is allowed
-                if (!is_overwrite_okay && pass == 0)
-                {
-                    d_keys = num_passes % 2 == 0 ?
-                        DoubleBuffer<KeyT>(d_keys_tmp, d_keys_tmp2) :
-                        DoubleBuffer<KeyT>(d_keys_tmp2, d_keys_tmp);
-                    d_values = num_passes % 2 == 0 ?
-                        DoubleBuffer<ValueT>(d_values_tmp, d_values_tmp2) :
-                        DoubleBuffer<ValueT>(d_values_tmp2, d_values_tmp);
-                }
-                d_keys.selector ^= 1;
-                d_values.selector ^= 1;
+                input_mode = output_is_tmp ? TMP_STORAGE : OUTPUT;
+                output_is_tmp = !output_is_tmp;
```

Comment:
@allisonvacanti Before I continue my work, I'd like to hear your feedback about this part. In order to support iterators, I have to add this verbose logic because the type of input iter, type of tmp storage, and type of output iter can be different.

Reply:
I am not sure if you are OK with this change, but this is very verbose, and I can not think of a better solution.

Reply:
We'll need to restore the `is_overwrite_okay` optimization first and then we can take a closer look at this part.